File: linked-lists/js/src/doubly-linked-list.js
class Node {
    next = null;
    prev = null;

    constructor(val) {
        this.value = val;
    }
}

export class DoublyLinkedList {
    head = null;
    tail = null;

    constructor() {
    }

    append(val) {
        const newNode = new Node(val);

        if (!this.tail) {
            this.tail = newNode;
        }

        if (!this.head) {
            this.head = newNode;
            return newNode;
        }

        const lastNode = this.tail;
        this.tail = newNode;
        this.tail.prev = lastNode;
        lastNode.next = this.tail;

        return newNode;
    }

    insertAt(node, val) {
        const newNode = new Node(val);
        newNode.next = node.next;
        node.next = newNode;

        return newNode;
    }

    traverse(fn) {
        let current = this.head;
        let last = null;

        while (current) {
            const result = fn(current, last);
            last = current;
            current = current.next;
            if (result) break;
        }

        return last;
    }

    traverseBackward(fn) {
        let current = this.tail;
        let first = null;

        while (current) {
            const result = fn(current, first);
            first = current;
            current = current.prev;
            if (result) break;
        }

        return first;
    }

    remove(node) {
        if (!this.head) return;

        // If the value found is the first in the list
        if (this.head.value === node.value) {
            this.head = this.head.next;
            return;
        }

        // If the value found is the last in the list
        if (this.tail.value === node.value) {
            this.tail = this.tail.prev;
            return;
        }

        this.traverse((cur, last) => {
            if (cur.value === node.value) {
                last.next = cur.next
                cur.next.prev = last
            }
        });
    }
}

File: linked-lists/go/pkg/double/doubly-linked-list.go
// Package double demonstrates Doubly Linked Lists
// https://dxt.rs/category/programming/general/linked-lists#doubly-linked-lists
package double

type Node struct {
	Val  string
	next *Node
	prev *Node
}

type DoublyLinkedList struct {
	head *Node
	tail *Node
}

type predicatefn func(*Node, *Node) bool

func (dl *DoublyLinkedList) Traverse(fn predicatefn) *Node {
	curr := dl.head
	var last *Node

	for curr != nil {
		res := fn(curr, last)
		last = curr
		curr = curr.next
		if res {
			break
		}
	}

	return last
}

func (dl *DoublyLinkedList) TraverseBackward(fn predicatefn) *Node {
	curr := dl.tail
	var first *Node

	for curr != nil {
		res := fn(curr, first)
		first = curr
		curr = curr.prev
		if res {
			break
		}
	}

	return first
}

func (dl *DoublyLinkedList) InsertAt(n *Node, v string) *Node {
	newNode := Node{
		Val:  v,
		next: n.next,
		prev: n,
	}
	n.next = &newNode

	if n == dl.tail {
		dl.tail = &newNode
	}

	return &newNode
}

func (dl *DoublyLinkedList) Append(v string) *Node {
	newNode := Node{
		Val: v,
	}

	if dl.head == nil {
		dl.head = &newNode
		return &newNode
	}

	lastNode := dl.Traverse(func(_, _ *Node) bool { return false })
	lastNode.next = &newNode
	newNode.prev = lastNode
	dl.tail = &newNode

	return &newNode
}

func (dl *DoublyLinkedList) Remove(n *Node) {
	if dl.head == nil {
		return
	}

	if dl.head.Val == n.Val {
		dl.head = dl.head.next
		return
	}

	dl.Traverse(func(nn, last *Node) bool {
		if nn.Val == n.Val {
			last.next = n.next

			if n == dl.tail {
				dl.tail = last
			}

			return true
		}
		return false
	})
}
